pkg/tempo/servicegraphprocessor/processor.go

```diff
@@ -0,0 +1,227 @@
+	serverService, clientService string
+	serverLatency, clientLatency time.Duration
+
+	failed bool
+}
+
+func (e *edge) complete() bool {
+	return len(e.clientService) != 0 && len(e.serverService) != 0
+}
+
+var _ component.TracesProcessor = (*processor)(nil)
+
+type processor struct {
+	nextConsumer consumer.Traces
+	reg          prometheus.Registerer
+
+	store    *cache.Cache
+	maxEdges int
+
+	serviceGraphRequestTotal       *prometheus.CounterVec
+	serviceGraphRequestFailedTotal *prometheus.CounterVec
+	serviceGraphRequestHistogram   *prometheus.HistogramVec
+
+	closed atomic.Bool
+
+	logger log.Logger
+}
+
+func newProcessor(nextConsumer consumer.Traces, cfg *Config) (*processor, error) {
+	if cfg.wait == 0 {
+		cfg.wait = defaultWait
+	}
+	if cfg.maxEdges == 0 {
+		cfg.maxEdges = defaultMaxEdges
+	}
+
+	p := &processor{
+		nextConsumer: nextConsumer,
+		store:        cache.New(cfg.wait, cfg.wait*2),
+		maxEdges:     defaultMaxEdges,
+		closed:       atomic.Bool{},
+	}
+
+	return p, nil
+}
+
+func (p *processor) Start(ctx context.Context, _ component.Host) error {
+	var reg prometheus.Registerer
+	reg = prometheus.NewRegistry()
+	if v, ok := ctx.Value(contextkeys.PrometheusRegisterer).(prometheus.Registerer); ok {
+		reg = v
+	}
+	p.reg = reg
+
+	return p.registerMetrics()
+}
+
+func (p *processor) registerMetrics() error {
+	p.serviceGraphRequestTotal = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "tempo_service_graph_request_total",
+		Help: "Total count of requests between two nodes",
+	}, []string{"client", "server"})
+	p.serviceGraphRequestFailedTotal = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "tempo_service_graph_request_failed_total",
+		Help: "Total count of failed requests between two nodes",
+	}, []string{"client", "server"})
+	p.serviceGraphRequestHistogram = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Name:    "tempo_service_graph_request_seconds",
+		Help:    "Time for a request between two nodes",
+		Buckets: prometheus.ExponentialBuckets(0.01, 2, 12),
+	}, []string{"client", "server"})
+
+	cs := []prometheus.Collector{
+		p.serviceGraphRequestTotal,
+		p.serviceGraphRequestFailedTotal,
+		p.serviceGraphRequestHistogram,
+	}
+
+	for _, c := range cs {
+		if err := p.reg.Register(c); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *processor) Shutdown(context.Context) error {
+	p.closed.Store(true)
+	p.unregisterMetrics()
+	return nil
+}
+
+func (p *processor) unregisterMetrics() {
+	cs := []prometheus.Collector{
+		p.serviceGraphRequestTotal,
+		p.serviceGraphRequestFailedTotal,
+		p.serviceGraphRequestHistogram,
+	}
+
+	for _, c := range cs {
+		p.reg.Unregister(c)
+	}
+}
+
+func (p *processor) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{}
+}
+
+func (p *processor) ConsumeTraces(ctx context.Context, td pdata.Traces) error {
+	if p.closed.Load() {
+		return nil
+	}
+
+	if p.store.ItemCount() >= p.maxEdges {
+		return ErrTooManyEdges
```

Comment:
unsure what returning err from this method does. logging/metricing this may make more sense. we don't want to interrupt the flow of spans if this processor is unable to generate a service graph.

Reply:
Changed to log the error and return nil to not drop spans